tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,372 @@
+) -> Optional[SymbolInformation]:
+    loc = Location(uri, location_to_range(decl.location()))
+    if kind := decl_kind(decl):
+        # TODO: should we use DocumentSymbol or SymbolInformation
+        # LSP spec says prefer DocumentSymbol, but nesting doesn't work out of the box. implies that we need some kind of visitor pattern to build a DS tree
+        # using symbol information for now, as it sort-of autogets the tree structure
+        is_deprecated = chapel.is_deprecated(decl)
+        return SymbolInformation(
+            loc, decl.name(), kind, deprecated=is_deprecated
+        )
+    return None
+
+
+@dataclass
+class NodeAndRange:
+    node: chapel.core.AstNode
+    rng: Range = field(init=False)
+
+    def __post_init__(self):
+        self.rng = location_to_range(self.node.location())
+
+    def get_location(self):
+        return Location(self.get_uri(), self.rng)
+
+    def get_uri(self):
+        path = self.node.location().path()
+        return f"file://{path}"
+
+
+@dataclass
+class ResolvedPair:
+    ident: NodeAndRange
+    resolved_to: NodeAndRange
+
+
+@dataclass
+class FileInfo:
+    uri: str
+    context: chapel.core.Context
+    segments: List[ResolvedPair] = field(default_factory=list)
+    siblings: chapel.SiblingMap = field(init=False)
+
+    def __post_init__(self):
+        self.rebuild_segments()
+
+    def parse_file(self) -> List[chapel.core.AstNode]:
+        """
+        Parses this file and returns the toplevel ast elements
+
+        Note: if there are errors they will be printed to the console.
+        This call should be wrapped an appropriate error context.
+        """
+
+        return self.context.parse(self.uri[len("file://") :])
+
+    def get_asts(self) -> List[chapel.core.AstNode]:
+        """
+        Returns toplevel ast elements. This method silences all errors.
+        """
+        with self.context.track_errors() as _:
+            return self.parse_file()
+
+    def rebuild_segments(self):
+        """
+        Rebuild the cached line info and siblings information
+
+        Note: this is a potentially expensive operation, it should only be done
+        when advancing the revision
+        """
+        asts = self.get_asts()
+        # get ids
+        self.segments = [
+            ResolvedPair(NodeAndRange(node), NodeAndRange(to))
+            for node, _ in chapel.each_matching(asts, chapel.core.Identifier)
+            if (to := node.to_node())
+        ]
+        self.segments.sort(key=lambda s: s.ident.rng.start)
+        self.siblings = chapel.SiblingMap(self.get_asts())
+
+    def get_segment_at_position(
+        self, position: Position
+    ) -> Optional[ResolvedPair]:
+        """lookup a segment based upon a Position, likely a user mouse location"""
+        idx = bisect_right(
+            self.segments, position, key=lambda s: s.ident.rng.start
+        )
+        idx -= 1
+        if not (idx >= 0 and position < self.segments[idx].ident.rng.end):
```

Comment:
Consider de-morganing this:

```suggestion
        if idx < 0 or position >= self.segments[idx].ident.rng.end:
```